mm2src/coins/z_coin.rs

```diff
@@ -1952,6 +2031,83 @@ impl InitWithdrawCoin for ZCoin {
     }
 }
 
+/// Waits until there are enough _unlocked_ Sapling notes to cover `total_required`.
+async fn wait_for_spendable_balance_impl(
+    selfi: ZCoin,
+    total_required: BigDecimal,
+) -> Result<impl Iterator<Item = SpendableNote>, MmError<GenTxError>> {
+    const MAX_RETRIES: usize = 40;
+    const RETRY_DELAY: f64 = 15.0;
+
+    let mut retries = 0;
+
+    loop {
+        let spendable_notes = selfi
+            .spendable_notes_ordered()
+            .await
+            .map_err(|e| GenTxError::SpendableNotesError(e.to_string()))?;
+        let spendable_notes_len = spendable_notes.len();
+
+        let locked_notes = selfi.z_fields.locked_notes_db.load_all_notes().await?;
+
+        let unlocked_notes: Vec<SpendableNote> = if locked_notes.is_empty() {
+            spendable_notes
+        } else {
+            let locked_seeds: HashSet<_> = locked_notes.into_iter().map(|n| n.rseed).collect();
+            spendable_notes
+                .into_iter()
+                .filter(|note| !locked_seeds.contains(&rseed_to_string(&note.rseed)))
+                .collect()
+        };
+
+        let sum_available: BigDecimal = unlocked_notes
+            .iter()
+            .map(|n| big_decimal_from_sat_unsigned(n.note_value.into(), selfi.decimals()))
+            .sum();
+
+        if sum_available >= total_required || unlocked_notes.len() == spendable_notes_len {
+            return Ok(unlocked_notes.into_iter());
+        }
```

Comment:
shouldn't we return an error if `unlocked_notes.len() == spendable_notes_len` when `sum_available` is less than the required? I know we handle this error in the caller but this means this function does 2 things and is not very clear, doesn't wrap it's logic in a neat way.

Reply:
> shouldn't we return an error if unlocked_notes.len() == spendable_notes_len when sum_available is less than the required? 

You're correct.

Simplified this a bit. Thanks for the catch.

Reply:
https://github.com/KomodoPlatform/komodo-defi-framework/blob/dd4d049638b78732ef6945da9040983bd3f2715b/mm2src/coins/z_coin.rs#L2060-L2073
The above can be written in a much simpler way
```rust
        if sum_available >= total_required {
            return Ok(unlocked_notes.into_iter());
        }

        if  unlocked_notes.len() == spendable_notes_len {
            // return error here
        }
```
This is what I meant

Reply:
This looks like a bug being introduced since we might have a locked note we should wait for..
```rust
 if  unlocked_notes.len() == spendable_notes_len {
            // return error here
   }
```
so we need a more explicit check like before returning error
```rust
if sum_available < total_required && unlocked_notes_len == spendable_notes_len
```

Reply:
but we already checked `if sum_available >= total_required` so `sum_available < total_required` is always true after that.

Reply:
this will lead to a bug if we have  let's say`1 locked_note`(we need to wait for unlock) and `1 spendable_note`
```rust
if  unlocked_notes.len() == spendable_notes_len {
    // if we return an error here without checking sum_available < total_required, then tx will fail unintended because we might have.
    We can only return an error here if unlocked_notes.len() == spendable_notes_len and sum_available < total_required
}
```